---
title: R-code for "Chemical signatures encode colony membership, mother-offspring similarity,
  relatedness and genetic quality in Fur seals""
author: Stoffel, M.A., Caspers, B.A., Forcada, J., Giannakara, A., Baier, M.C., Eberhart-Phillips,
  L.J. , Müller, C. & Hoffman, J.I.
output:
  pdf_document: default
  html_document:
    highlight: pygments
bibliography: cites.bib
---
```{r, echo = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>", cache = TRUE) # warning = FALSE
```

This document provides the code for all major analysis from our paper. Duplicate analyses that have not been part of an argument (e.g. most analysis for pups) were strapped out for readability. All supplemented data files should be placed in a subfolder called "files" for correct loading and functionality of the code.
The Rmarkdown file as well as the data are stored on GitHub. For any questions just contact me: `martin.adam.stoffel@gmail.com`
The sequence of code follows roughly the sequence of analysis in the *Results* section of the paper.

**Prerequisites:**

* We wrote a package for doing some of the inbreeding related analysis, such as calculating g2 or sMLH which is hosted on GitHub. The `inbreedR` package provides functions for measuring inbreeding from molecular data (SNPs and microsatellites) and will soon be published. To download packages from GitHub repositories, you need to install the `devtools` package. 
* For running the complete code you need  a `files` subfolder with all the raw data files.


Installing `inbreedR` from GitHub.

```{r, eval = FALSE}
# install.packages("devtools")
library(devtools)
install_github("mastoffel/inbreedR")
```

```{r, results="hide", message=FALSE, cache = TRUE}
library(devtools)
library(inbreedR)
```

See `?inbreedR` for further information on the functions. 

***
## Loading data, standardisation and transformation

Loading the 

- **chemical data** (scent_raw, called scent data from now), which is the output of Gas-chromatography peak detection was done in Xcalibur 2.0.5. (A first preprocessing was done by aligning the raw chemical data and removing substances that have been present in the control sample, see Methods part of the paper)*
- **identities** for colony membership (colony), mother-offspring pairs (family) and mothers and pups, respectively (age)

```{r}
scent_raw <- as.data.frame(t(read.csv(".\\files\\scent_raw.csv", row.names = 1)))
factors <- read.csv(".\\files\\factors.csv",row.names=1) 
head(factors)
```

Standardising observations by total, such that within every observation compounds
add up to 100 % (Thus averaging out absolute concentration differences between samples)

```{r}
scent_stand <- as.data.frame(t(apply(scent_raw, 1, function(x) (x/sum(x)) * 100)))
```

Log(x+1) transformation of the standardised scent data.

```{r}
scent <- log(scent_stand + 1)
```

The scent matrix contains 82 observations and 213 compounds (retention times of chemicals are column names, values are relative concentrations) in total.

```{r, fig.width = 5, fig.height = 5}
dim(scent)
head(scent[1:6])
```

***

## Colony differences in chemical fingerprints

```{r, results="hide", message=FALSE}
library(vegan)
library(MASS)
```  
    
**Non-metric multidimensional scaling (nMDS)** visualizes a distance matrix (Bray-Curtis similarity).
The nMDS algorithm aims to place each individual in a 2-dimensional space such that the
between-individual distances are preserved as well as possible. Axis coordinates are 
arbitrary and not shown. The plot is better visualized with colours (see paper) and
is shown here for the purpose of demonstration. Mother-offspring pairs can be
identified by labels (e.g. M14, P14).

```{r, results = FALSE}
scent_mds <- MASS::isoMDS(vegdist(scent))
```

```{r, fig.height=6, fig.width=6, warning = FALSE, results = FALSE}
vegan::ordiplot(scent_mds, type = "t", ylab = "", xlab = "",axes=FALSE, frame.plot=TRUE)
```

Analysis of Similarities (ANOSIM) is a non-parametric test for group differences based on a Bray-curtis (or any other) similarity matrix. We use the vegan package [@vegan] for ANOSIM and several other functions. Most analysis are done for the whole sample as well as for mothers and pups seperately to avoid pseudoreplication. ANOSIM is based on a permutation test, which is why results can slightly differ from the paper.

**Dissimilarity between the two colonies.**
```{r, cache=TRUE}
vegan::anosim(dat = scent, grouping = factors$colony, 
       distance = "bray", permutations = 1000)
```

**Dissimilarity between mothers from the two colonies.**
```{r, cache=TRUE}
vegan::anosim(dat = scent[factors$age == 1, ], grouping = factors$colony, 
       distance = "bray", permutations = 1000)
```

**Dissimilarity between pups from the two colonies.**
```{r, cache=TRUE}
vegan::anosim(dat = scent[factors$age == 2, ], grouping = factors$colony, 
       distance = "bray", permutations = 1000)
```

*Genetic differentiation of the two colonies was assesed through bayesian structure analysis, with the software "Structure" [@Pritchard:2000uv]*  

***

## Mother offspring similarity in chemical fingerprints. 

**Full sample**
```{r, cache=TRUE}
vegan::anosim(dat = scent, grouping = factors$family, 
       distance = "bray", permutations = 1000)
```

**Mother offspring similarity within colony 1 (Special study beach)**
```{r, cache=TRUE}

vegan::anosim(dat = scent[factors$colony == 1, ], 
       grouping = factors[factors$colony == 1, ]$family, 
       distance = "bray", permutations = 1000)
```

**Mother offspring similarity within colony 2 (Freshwater beach)**
```{r, cache=TRUE}
vegan::anosim(dat = scent[factors$colony == 2, ], 
       grouping = factors[factors$colony == 2, ]$family, 
       distance = "bray", permutations = 1000)
```
  

***

## Chemical similarity vs. geographic distance on special study beach 

* location data in meters is available for this population as the special study beach on Bird Island provides an aerial walkway
 
Loading X-Y coordinates of each individual.

```{r}
coord  <- read.csv(".\\files\\coordinates_beach1.csv", row.names=1) 
head(coord)
```

Converting coordinates to pairwise euclidian distance matrix.

```{r}
dist_mat <- as.matrix(dist(coord, method = "euclidian"))
```

Constructing a bray curtis similarity matrix (from chemical fingerprints) of all individuals from beach 1 (special study beach). We constantly used spearman rank correlation in mantel tests.

```{r}
scent_bc <- as.matrix(vegan::vegdist(as.matrix(scent[factors$colony == 1, ])),
                      method = "bray")
```

**Geographic distance vs. chemical similarity in mothers**
```{r}
geo_mum <- dist_mat[1:20, 1:20]
scent_mum <- scent_bc[1:20, 1:20]
vegan::mantel(geo_mum, scent_mum, method = "spearman")
```

**Geographic distance vs. chemical similarity in pups**
```{r}
geo_pup <- dist_mat[21:40, 21:40]
scent_pup <- scent_bc[21:40, 21:40]
vegan::mantel(geo_pup, scent_pup, method = "spearman")
```

***

## Correlation between genotype and overall chemical fingerprints. 

#### Relatedness and overall chemical similarity

Load pairwise relatedness [@Queller:1989wl] based on 41 microsatellite markers.
```{r, fig.width = 5, fig.height = 4}
relatedness <- as.matrix(read.csv(".\\files\\relatedness.csv",row.names=1))
head(relatedness[1:6, 1:6])
hist(relatedness)
```

**Pairwise bray curtis similarity in chemical fingerprints of all individuals.**
```{r}
scent_bc <- 1-(as.matrix(vegan::vegdist(as.matrix(scent)), method = "bray")) 
head(scent_bc[1:6, 1:6])
```

**Mantel test between genetic relatedness and bray curtis similarity in chemical fingerprints of all individuals.** 
```{r}
vegan::mantel(relatedness, scent_bc, method = "spearman", permutation = 1000)
```

We find a significant relationship between the overall chemical fingerprints and genetic relatedness. However, we are likely to have a problem of pseudoreplication here. For that reason, we are analysing mothers and pups seperately.    
  
**Fur seal mothers: mantel test between  genetic relatedness and bray curtis similarity of chemical  fingerprints.**

```{r}
vegan::mantel(relatedness[factors$age == 1, factors$age == 1], 
              scent_bc[factors$age == 1, factors$age == 1],
              method = "spearman", permutation = 1000)
```

**Fur seal pups: mantel test between genetic relatedness and bray curtis similarity of chemical fingerprints.**

```{r}
vegan::mantel(relatedness[factors$age == 2, factors$age == 2], 
              scent_bc[factors$age == 2, factors$age == 2],
              method = "spearman", permutation = 1000)
```

***
## Correlation between heterozygosity (sMLH) and diversity (number of compounds) of chemical fingerprints

* The function `sMLH`  is part of the `inbreedR` package, currently available on GitHub. 

```{r}
library(inbreedR)
# ?inbreedR
```

Loading raw genotypes and calculating standardised multilocus heterozygosity (sMLH) based on 41 markers.
* `inbreedR`package requires a special format, see `?convert_raw` for more information`*

```{r}
genotypes <- read.table(".\\files\\genotypes.txt", row.names=1)
genotypes[1:6, 1:6]
genotypes_formatted <- inbreedR::convert_raw(genotypes, miss_val = NA)  
heterozygosity <- inbreedR::sMLH(genotypes_formatted)
```

Number of compounds per individual.

```{r}
num_comp <- as.vector(apply(scent, 1, function(x) length(x[x>0])))
```

**Linear model of heterozygosity on number of compounds in mothers**

A clear association between sMLH and chemical complexity in mothers but not pups.

```{r}
het_mum <- heterozygosity[factors$age == 1]
num_comp_mum <- num_comp[factors$age==1]
summary(lm(het_mum ~ num_comp_mum))
```

Plotting with ggplot2, an implementation of the grammar of graphics [@ggplot2]

```{r, echo = FALSE, fig.width=6, fig.height=5}
het_df <- as.data.frame(cbind(het_mum, num_comp_mum))
library(ggplot2)
library(grid)
ggplot2::ggplot(het_df, aes(x=het_mum, y=num_comp_mum)) +
        geom_point(colour = "black", size = 2.5) +
        geom_smooth(method="lm",size = 1 ,alpha=0.13, colour="black", fill = "lightblue") +
        theme_minimal(base_size = 14) +
        theme(strip.text.x = element_text(vjust=1,size = 14),
              axis.title.x = element_text(vjust= -2 ,size = 14),
              axis.title.y = element_text(vjust=3,size = 14),
              axis.ticks.x = element_blank(),
              axis.ticks.y = element_blank(),
              plot.margin = (unit(c(.5, .5, 2, 2), "cm"))) +
        scale_x_continuous(breaks=c(seq(from = 0.8, to = 1.20, by = 0.1))) +
        geom_text(aes(0.85,50, label="r = 0.34, p = 0.027"),size=4, colour = "blue") +
        labs(title = "Correlation of sMLH and chemical complexity \nin Fur seal mothers")+
        xlab("sMLH") +
        ylab("number of compounds") 
```

**Linear model of heterozygosity on number of compounds in pups**

```{r}
het_pup <- heterozygosity[factors$age == 2]
num_comp_pup <- num_comp[factors$age==2]
summary(lm(het_pup ~ num_comp_pup))
```


## Strength of correlation between sMLH and number of compounds increases with an increasing number of genetic markers in mothers.

The `resample_loci()` function samples an increasing subset of loci, calculates sMLH and correlates with a vector y (here: number of compounds in chemical fingerprints). 

```{r, cache = TRUE}
resample_loci <- function(genotypes, y, num_iter = 1000) {
# genotypes in inbreedR format. See ?inbreedR
# y is a vector to correlate with sMLH
# num_iter is the number of resamplings per added locus
        # calculate number of loci
        num_loci <- ncol(genotypes)
        results <- data.frame(matrix(nrow = num_iter, ncol = num_loci))
        for (i in seq_along((1: num_loci))){
                for (k in seq_along(1:num_iter)) {
                loci_ind <- sample(1:num_loci, i, replace = FALSE)
                het <- inbreedR::sMLH(genotypes[, loci_ind])
                results[k, i] <- cor(het[1:41],y) # heterozygosity subsetted for mothers
                }
        }
results
}

# Converting genotypes into the right format 
genotypes_formatted <- inbreedR::convert_raw(genotypes, miss_val = NA)  
# Resampling 1 - 40 loci each 1000 times, compute sMLH and correlate with number of compounds
resample_mums <- resample_loci(genotypes_formatted, num_comp_mum, num_iter = 1000)
```

Calculating summary statistics for the resampling output: mean, sd, se of the correlations per subset of markers.

```{r}
sum_results <- function(resampling_output) {
        mean_cor <- apply(resampling_output,2,mean, na.rm=T)
        sd_cor <- apply(resampling_output,2,sd, na.rm=T)
        se_cor <- sd_cor/(sqrt(nrow(resampling_output)))
        sum_results <- data.frame(locnum = 1:ncol(resampling_output), 
                                  cormean = mean_cor, corsd = sd_cor, corse = se_cor)
}

results_mums <- sum_results(resample_mums) 
```

**Plotting mean correlation of heterozygosity (estimated by an increasing number of markers) with number of compounds in chemical fingerprints for Fur seal mothers.**

*Pups are not shown here for simplicity and to avoid code replication. For the full figure see the results section of the paper*

```{r, fig.width=6, fig.height=5}
# plotting
library(grid)
ggplot2::ggplot(results_mums, aes(x = locnum, y = cormean)) +
        geom_line(size = 0.6, colour = "black") +
        geom_errorbar(aes(ymin = cormean-corse, ymax = cormean+corse),
                      width=0.8, alpha=0.7, size = 0.8, colour = "black") +
        geom_point(size = 2, shape = 16) +
        theme_minimal(base_size = 16) +
        theme(axis.title.x = element_text(vjust= -2 ,size = 16),
              axis.title.y = element_text(vjust=3,size = 16),
              axis.ticks.x = element_blank(),
              axis.ticks.y = element_blank(),
              plot.margin = (unit(c(.5, .5, 2, 2), "cm"))) +
        #geom_hline(yintercept=0.305) +
        ylab("r (sMLH, ncomp) +- se") +
        xlab("number of loci") +
        labs(title = "r between number of compounds \nand sMLH estimated \nfrom an increasing number of loci")
```


## Estimation of identity disequilibrium g2 with the inbreedR package. (can diverge slightly from the RMES program)

Instead to just finding a correlation between heterozygosity and a trait such as chemical complexity, one can ask whether variation in inbreeding (so called-general effects) is a potential cause. This can be measured with a parameter called g2 [@david2007reliable], that assesses identity disequilibrium through quantification of excess double heterozygote loci. 
We are currently working on the inbreedR package, which provides functions for calculation g2 with both microsatellites and SNPs.

Calculate g2.
```{r, cache=TRUE, results = "hide"}
g2 <- inbreedR::g2_microsats(genotypes_formatted, nperm = 1000, nboot = 1000, CI = 0.95)
```

```{r, cache=TRUE, echo = FALSE}
g2
```
*potentially make figure for increasing markers here*

***

## Factor analysis on the chemical compounds data with the package HDMD.

HDMD [@HDMD] allows for doing a Factor analysis with high dimensional data(where the number of variables exceeds the number of observations) by calculating a general inverse matrix. 
```{r}
library(HDMD)
library(minmodelr)
source("get_pairdiff.R") 
```

Factor analysis and extraction of factor scores for the first 4 factors. Promax rotation of the factors allows them to be non-orthogonal and thus correlated. After FA, the factor scores for each individual on all 4 factors are extracted.

```{r}
# factor analysis with 4 factors, promax rotation ------------------------------
scent_fa <- HDMD::factor.pa.ginv(scent, nfactors = 4, 
                           prerotate = T,rotate = "promax", 
                           scores = T, m = 3)
fa_scores <- as.data.frame(scent_fa$scores)
head(fa_scores)
```

The eigenvalue course seen in the screeplot allows for decisions on the number of factors to retain.

```{r, fig.width=6, fig.height=5}
# screeplot, 4 factors left to the "scree"
plot(scent_fa$values[1:8], type="b", ylab = "eigenvalue", xlab = "factor", 
     main = "Screeplot")
```

Plotting the distribution of factor scores seperately for each colony. Similar distributions suggest the compounds which are represented by a given factor to be similarly distributed across colonies and could thus be of potential genetic origin, while different distributions as in factor 4 suggest this factor to represent environmentally influenced compounds.

```{r}
# distribution of factor scores
df <- cbind(fa_scores, factors["colony"])
df$colony <- as.factor(df$colony)

for (i in c(1,2,4)) {
plot_all <- ggplot(df, aes_string(x = paste("F", i, sep = ""), fill = "colony")) +
        geom_density(alpha=0.8, size=0.5, aes(fill = colony),adjust=1.5) +
        scale_fill_manual(values = c("blue","red")) +
        guides(fill=guide_legend(title=NULL)) +
        theme_minimal(base_size = 16) + 
        theme(legend.position="none") +
        scale_x_continuous(breaks = c(seq(from = -1, to = 6, by = 1))) + 
        scale_y_continuous(breaks = c(seq(from = 0, to = 1.4, by = 0.4))) +  
        xlab(paste("Factor", i, sep = " ")) +
        ylab("Density")

assign(paste("f", i, "_plot", sep = ""), plot_all)
}
```


**Multiplot function from `cookbook-r.com` for plotting multiple ggplots**
```{r}
multiplot <- function(..., plotlist=NULL, cols) {
  require(grid)
  
  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)
  
  numPlots = length(plots)
  
  # Make the panel
  plotCols = cols                          # Number of columns of plots
  plotRows = ceiling(numPlots/plotCols) # Number of rows needed, calculated from # of cols
  
  # Set up the page
  grid.newpage()
  pushViewport(viewport(layout = grid.layout(plotRows, plotCols)))
  vplayout <- function(x, y)
    viewport(layout.pos.row = x, layout.pos.col = y)
  
  # Make each plot, in the correct location
  for (i in 1:numPlots) {
    curRow = ceiling(i/plotCols)
    curCol = (i-1) %% plotCols + 1
    print(plots[[i]], vp = vplayout(curRow, curCol ))
  }
  
}
```

Plotting all factor distributions.
```{r}
multiplot(f1_plot, f2_plot, f4_plot, cols = 2)
```
  
**Linear model of heterozygosity on factors (factor scores) as explanatory variables in mothers.** 

```{r}
# bind heterozygosity and the factor scores in one data.frame and subset mothers
het_df <- cbind(heterozygosity, fa_scores)[factors$age == 1, ]
het_model <- lm(heterozygosity ~., data=het_df)
summary(het_model)
```

While Factor1 and Factor2 seem to represent substances that are accociated with heterozygosity, Factor 3 and Factor 4 clearly don´t. To simplify the model we used deletion testing (Crawley, Statistics). The minmodelr package contains some helper functions for this task. See ?MinMod, ?DelTestVar. *We don´t generally recommend a deletion testing procedure. In our case, results are clear and we use it for simplicity rather than for fishing significant results.*

```{r}
library(devtools)
# install_github("mastoffel/minmodelr")
library(minmodelr)
```

```{r}
het_reduced <- minmodelr::MinMod(het_df)
# extract data frame
het_reduced_df <- het_reduced[[1]]
# extract reduced model
het_reduced_mod <- het_reduced[[2]]
# deletion testing for both variables in the reduced model. See ?DelTestVar
table <- minmodelr::DelTestVar(het_reduced_df)
# deviance explained by the reduced model
dev_expl <- (het_reduced_mod$null.deviance - het_reduced_mod$deviance) / het_reduced_mod$null.deviance
summary(het_reduced_mod)
```
  
Creating a new variable F1F2 which is the sum of the two factor scores and using this variable as predictor in a linear model of heterozygosity.
```{r}
# sum of factors as variable
het_df$F1F2 <- het_df$F1 + het_df$F2
table <- minmodelr::DelTestVar(as.data.frame(cbind(het_df$heterozygosity, het_df$F1F2)))
summary(lm(heterozygosity ~ F1F2, data = het_df))
```

#### Linear model of genetic relatedness on factor scores as explanatory variables for mothers.  

Pairwise genetic relatedness is represented as a matrix. To model the relationship between relatedness and factor scores we created a matrix for each factor, whereby each pairwise value represents the difference in factor scores for a pair of seals.

`get_pairdiff()` creates these matrices. We based these analysis on mothers and pups seperately.

```{r}
get_pairdiff <- function(relate, scores, df=F) {
# creates data.frame with  
# pairwise differences in factor scores
# input should be: relatedness data frame (lower triangular), 
# data frame with factor scores in columns
# if df=TRUE, get_pairdiff will return a list of score-difference
# dataframes (for each component/factor) with pairwise pc-differences. 
# make sure to have data.frames
relate <- as.data.frame(relate)
scores <- as.data.frame(scores)
# copy similarity matrix and clear
score_mat <- relate
score_mat[, ] <- NA
# get vector of pairwise-rownames
allnames <- vector()
for (i in 1:ncol(relate)) {
        for (k in 1:nrow(relate)) {
                nametemp <- paste(names(relate)[i], 
                                  row.names(relate)[k], sep = "")
                allnames <- append(allnames, nametemp)
        }
}
# roll out as vector
relate_vec <- unlist(relate)
# label the rows
names(relate_vec) <- allnames
# delete na´s
relate_vec <- relate_vec[!is.na(relate_vec)]
# get new row-names vector
pairnamessub <- names(relate_vec)
# create raw data frame 
fac_diff_all <- data.frame("relatedness"= relate_vec)
# construct similarity matrix out of pairwise differences in factors
names <- rownames(relate)
row.names(scores) <- names
fac_diff_mats <- list()

for (z in 1:ncol(scores)) {
        for (i in names) {
                for (k in names) {
                        if (!(is.na(relate[i,k]))) {
                                diff_fac <- abs(scores[i,z] - scores[k,z])
                                score_mat[i,k] <- diff_fac
                        }
                }
        }
        
        # create list of data frames, containing difference matrices per Factor
        fac_diff_mats <- c(fac_diff_mats, list(score_mat))
        
        # turn into vector
        factor_diff <- as.vector(as.matrix(score_mat))
        factor_diff <- factor_diff[!is.na(factor_diff)]
        fac_diff_all <- cbind(fac_diff_all, factor_diff)
}
## check argument for what to return
if (df == T) {
        names(fac_diff_mats) <- names(scores)
        return(fac_diff_mats)
} else if (df == F) {
        names(fac_diff_all) <- c("relatedness",names(scores))
        row.names(fac_diff_all) <- pairnamessub
        return(fac_diff_all)
}
}
```


Creation of 4 pairwise distance matrices for each factor.
```{r}
# source("get_pairdiff.R")
fa_diff_mums <- get_pairdiff(relatedness[factors$age == 1, factors$age == 1],
                             fa_scores[factors$age == 1, ], df = F)

# assign pairwise difference factor matrices to names
for (i in seq_along(1:4)) {
        assign(paste("f", i, "_diff", sep=""), fa_diff_mums[, i+1])
}
```

The ecodist package [@ecodist] can handle multiple distance matrices by doing a partial mantel test. 

*Every partial mantel test just tests for the association with the first response, while the other are permutated*

```{r}
rel_dist <- as.dist(relatedness[factors$age == 1, factors$age == 1])
ecodist::mantel(rel_dist ~ f1_diff + f2_diff + f3_diff + f4_diff, mrank = T, nperm = 1000)
ecodist::mantel(rel_dist ~ f2_diff + f1_diff + f3_diff + f4_diff, mrank = T)               
ecodist::mantel(rel_dist ~ f3_diff + f2_diff + f1_diff + f4_diff, mrank = T)               
ecodist::mantel(rel_dist ~ f4_diff + f3_diff + f2_diff + f1_diff, mrank = T)               
```

#### Linear model for associations between genetic relatedness and factor scores as explanatory variables for pups.
```{r}
fa_diff_pups <- get_pairdiff(relatedness[factors$age == 2, factors$age == 2],
                             fa_scores[factors$age == 2, ], df = F)

for (i in seq_along(1:4)) {
        assign(paste("f", i, "_diff", sep=""), fa_diff_pups[, i+1])
}

rel_dist <- as.dist(relatedness[factors$age == 2, factors$age == 2])
ecodist::mantel(rel_dist ~ f1_diff + f2_diff + f3_diff + f4_diff, mrank = T)               
ecodist::mantel(rel_dist ~ f2_diff + f1_diff + f3_diff + f4_diff, mrank = T)           
ecodist::mantel(rel_dist ~ f3_diff + f2_diff + f1_diff + f4_diff, mrank = T)                
ecodist::mantel(rel_dist ~ f4_diff + f3_diff + f2_diff + f1_diff, mrank = T)              
```

#### Colony differences in factor scores: Just factor 4 shows significant differences.
```{r}
col_df <- cbind(factors["colony"], fa_scores)
col_reduced <- minmodelr::MinMod(col_df)
col_reduced_df <- col_reduced[[1]]
# dev_expl <- (col_reduced_df$null.deviance - col_reduced_df$deviance) / col_reduced_df$null.deviance
table <- minmodelr::DelTestVar(col_reduced[[1]])
```

## Identification of substance subsets.
```{r}
# subsets and identification
library(vegan)
library(ggplot2)
library(dplyr)
library(magrittr)
library(vegan)
library(reshape2)
```

Similarity percentages analysis (`simper`) identifies the contribution of a specific compound to group similarity / dissimilarity. ANOSIM was used to test whether a small subset of the compounds with the highest contributions shows significant patterns.

#### Identification of best substances encoding mother-offspring similarity. 
For this analysis we have 41 groups (mother-offspring pairs) and want to look at withing group similarities rather then between group dissimilarities. This was done in Primer-E, as the `simper` function from the vegan package computes discriminating compounds, rather then compounds that make a mother-pup pair unique (although both sets overlap of course) .
```{r}
# results from simper analysis in Primer-E
mp_simp <- read.csv(".\\files\\simper_mp_results.csv", colClasses = c("character", "numeric"))
mp_simp
```

**Mother offspring similarity based on a Bray-curtis similarity matrix which was computed from just the subset of 12 top compounds from the SIMPER analysis is highly significant, both overall, as well as within colonies.**

**Full sample**

```{r}
vegan::anosim(dat = scent[mp_simp$comp], grouping = factors$family, 
       distance = "bray", permutations = 1000)
```

**Within colony 1 (Special study beach)**

```{r}
vegan::anosim(dat = scent[factors$colony == 1, mp_simp$comp], 
       grouping = factors[factors$colony == 1, ]$family, 
       distance = "bray", permutations = 1000)
```

**Within colony 2 (Freshwater beach)**

```{r}
vegan::anosim(dat = scent[factors$colony == 2, mp_simp$comp], 
       grouping = factors[factors$colony == 2, ]$family, 
       distance = "bray", permutations = 1000)
```

#### Identification of best substances encoding colony dissimilarity.

Using `simper` from the `vegan` package to find the important substances for discriminating between the two colonies. And sorting them subsequently in order of contribution to colony dissimilarity.
```{r}
# simper analysis
simp_colony <- vegan::simper(scent, factors$colony)

# getting 15 best substances and their contribution to colony dissimilarity
simp_colony_names <- rownames(summary(simp_colony, ordered = TRUE)[[1]])[1:15]
contribution <- summary(simp_colony, ordered = TRUE)[[1]]$contr[1:15]

# indices of colony substances (58,62,68,74,86,89,90,98,106,107,110,164,181,189,211)
ind_col <- paste(which(names(scent)%in%simp_colony_names), collapse = ",")

# connect to data frame and compute contribution in percent
col_simp <- data.frame(comp = simp_colony_names, contrib = contribution*100, stringsAsFactors = FALSE)
col_simp
```

**Colony dissimilarity based on 15 compounds.**
```{r}
# overall (number of permutations is 1000 instead of 10,000 in the paper)
anosim(dat = scent[col_simp$comp], grouping = factors$colony, 
       distance = "bray", permutations = 1000)
```

## Identification of substanced encoding relatedness.
All the following analyses are shown for the subset of mothers. 

The core of the idea is to use a bootstrapping procedure on the `BIO-ENV` function, originally by Clarke [@clarke2001primer], which was modified [@sinkr] to work with a bray curtis similarity matrix. For details see the methods part of the paper. The function is built to run on parallel with snowfall [@snowfall] on a server or similar, but still takes a couple of days to finish.

Additional packages used are Hadley Wickham´s dplyr [@dplyr] and stringr [@stringr].
First, the BIO-ENV function code is presented, followed by the bootstrap function.

```{r}
bio.env <- function(fix.mat, var.mat, 
                    fix.dist.method="bray", var.dist.method="euclidean",
                    scale.fix=FALSE, scale.var=TRUE,
                    output.best=10,
                    var.max=ncol(var.mat)
){
        # if(dim(fix.mat)[1] != dim(var.mat)[1]){stop("fixed and variable matrices must have the same number of rows")}
        if(var.max > dim(var.mat)[2]){stop("var.max cannot be larger than the number of variables (columns) in var.mat")}
        
        require(vegan)
        
        combn.sum <- sum(factorial(ncol(var.mat))/(factorial(1:var.max)*factorial(ncol(var.mat)-1:var.max)))
        
        if(scale.fix){fix.mat<-scale(fix.mat)}else{fix.mat<-fix.mat}
        if(scale.var){var.mat<-scale(var.mat)}else{var.mat<-var.mat}
        # fix.dist <- vegdist(fix.mat, method=fix.dist.method)
        fix.dist <- fix.mat
        RES_TOT <- c()
        best.i.comb <- c()
        iter <- 0
        for(i in 1:var.max){
                var.comb <- combn(1:ncol(var.mat), i, simplify=FALSE)
                RES <- data.frame(var.incl=rep(NA, length(var.comb)), n.var=i, rho=0)
                for(f in 1:length(var.comb)){
                        iter <- iter+1
                        var.dist <- vegdist(as.matrix(var.mat[,var.comb[[f]]]), method=var.dist.method)
                        temp <- suppressWarnings(cor.test(fix.dist, var.dist, method="spearman"))
                        RES$var.incl[f] <- paste(var.comb[[f]], collapse=",")
                        RES$rho[f] <- temp$estimate
                        if(iter %% 100 == 0){print(paste(round(iter/combn.sum*100, 3), "% finished"))}
                }
                
                order.rho <- order(RES$rho, decreasing=TRUE)
                best.i.comb <- c(best.i.comb, RES$var.incl[order.rho[1]])
                if(length(order.rho) > output.best){
                        RES_TOT <- rbind(RES_TOT, RES[order.rho[1:output.best],])
                } else {
                        RES_TOT <- rbind(RES_TOT, RES)
                }
        }
        rownames(RES_TOT)<-NULL
        
        if(dim(RES_TOT)[1] > output.best){
                order.by.best <- order(RES_TOT$rho, decreasing=TRUE)[1:output.best]
        } else {
                order.by.best <- order(RES_TOT$rho, decreasing=TRUE)
        }
        OBB <- RES_TOT[order.by.best,]
        rownames(OBB) <- NULL
        
        order.by.i.comb <- match(best.i.comb, RES_TOT$var.incl)
        OBC <- RES_TOT[order.by.i.comb,]
        rownames(OBC) <- NULL
        
        out <- list(
                order.by.best=OBB,
                order.by.i.comb=OBC,
                best.model.vars=paste(colnames(var.mat)[as.numeric(unlist(strsplit(OBB$var.incl[1], ",")))], collapse=",") ,
                best.model.rho=OBB$rho[1]
        )
        out
}
```



```{r, eval = FALSE}
################ BIO-ENV bootstrap procedure ###################################
#### run seperately on multicore server #####
#### aim: resampling test for finding the substances associated with genetic 
#### relatedness. Basic assumption: Each variable will be tested in many different 
#### environments (individuals, other variables), which will prevent spurious 
#### correlations, as the really important substances will occur in best subsets
#### in many different constellations. (see methods section)

# parallel computing using 40 cores, takes some days nevertheless and is just
# shown here.

library(vegan)
library(stringr)
library(dplyr)
library(snow)
library(snowfall)

# number of cores
ncores <- 2
# subset
scent_mum <- filter(scent, factors$age == 1)
relate_mum <-  relatedness[factors$age == 1, factors$age == 1]

# initialise results vector
all_best <- vector()

# initialise cluster
snowfall::sfInit(parallel=TRUE, cpus=ncores, type="SOCK")

# export libraries and main function to all cores
snowfall::sfSource("bio.env.R")
snowfall::sfLibrary(vegan)
snowfall::sfLibrary(stringr)
snowfall::sfLibrary(dplyr)

bootstrap <- function(iter_comp) { # main resampling function
        for (i in 1:500) {
                # sample 20 out of 41 mothers, indices
                ind_obs <- sort(sample(1:41, size = 20, replace = F))
                # subset relate_mum and scent_mum
                reltemp <- 1-as.dist(relate_mum[ind_obs, ind_obs])
                abundtemp <- scent_mum[ind_obs, ]
                for (i in iter_comp) {
                        # sample 10 compounds
                        index_comps <- sort(sample(1:213, size = 10, replace = F))
                        abundtemp_sub <- abundtemp[, index_comps]
                        # get vector with 0 for null-column and 1 for non-null column
                        nullcomps <- apply(abundtemp_sub, 2, function(x) sum(x>0))
                        abundtemp_sub <- subset(abundtemp_sub, 
                                                subset = c(rep(TRUE,nrow(abundtemp_sub))), 
                                                select = (nullcomps >= 2))
                        # new iteration if too less substances left
                        if (ncol(abundtemp_sub) <= 2) next
                        # main function: bio.env finds subset that mostly correlates
                        # with relatedness
                        results <- bio.env(reltemp, abundtemp_sub, 
                                           var.dist.method = "bray", 
                                           scale.fix = F, scale.var = F)
                        mods <- results$best.model.vars
                        best <- unlist(str_split(mods, ","))
                        all_best <- append(all_best, best)
                        # write(best, file = "best.txt", append = TRUE, sep = " ")
                }
        }
        return(all_best)
}
# export objects
snowfall::sfExportAll(except = NULL, debug = FALSE)
snowfall::sfClusterEval(ls())
# create list of 500 iterations for all cores
vals <- list()
for (i in 1:ncores) {
        vals[[i]] <- 1:500
}
# run analysis
# best is a list of all best subsets
best <- snowfall::sfLapply(vals, bootstrap)
# stop cluster
sfStop()
# bring all results 
results <- unlist(best)

############################## END #############################################
```

**Analysing results from the BIO-ENV bootstrap analysis.**

`best_mums` is a data frame containing the number of occurences of each variable in the best subset from the BIO-ENV bootstrap analysis. Substances, that were retained more often are therefore likely to be genuinly associated with genetic relatedness.
```{r}
# substance occurences are sorted in the table
best_mums <- read.csv("files/bootstrap_mums.csv",row.names=1)
```

To analyse how many of these compounds are really important, the idea is to take an increasing number of "best" compounds and compute a mantel test with relatedness for each of the subsets. The subsequent plot shows a nice peak, which could be seen as the optimal number of chemicals encoding relatedness.

```{r, warning=FALSE}
# subset mothers
scent_mum <- dplyr::filter(scent, factors$age == 1)
relate_mum <-  1-relatedness[factors$age == 1, factors$age == 1]

sub_names_mums <- row.names(best_mums)

statm <- vector()
sigm <- vector()

# compute mantelR for an increasing set of best substances
for (i in 2:100) {
        bc_dist <- vegan::vegdist(scent_mum[, sub_names_mums[1:i]], method = "bray")
        mod <- vegan::mantel(relate_mum, bc_dist, na.rm = T, method = "spearman")
        statm <- append(statm, mod$statistic)
        sigm <- append(sigm, mod$sig)
}

stat_df <- data.frame(num_comps = 1:length(statm), mantelR = statm)
```

**Plotting mantelR for an increasing number of best substances.**
```{r, fig.height=5, fig.width=6}
library(grid)
# simple plot
ggplot2::ggplot(stat_df, aes(x = num_comps, y = mantelR)) +
        stat_smooth(se = FALSE, span = 0.16, size = 1.3, method = "loess") +
        geom_point(colour = "black", size = 3, alpha = 0.8) +
        theme_minimal(base_size = 16) +
        theme(strip.text.x = element_text(vjust=1,size = 16),
              axis.title.x = element_text(vjust= -2 ,size =16),
              axis.title.y = element_text(vjust=3,size = 16),
              axis.ticks.x = element_blank(),
              axis.ticks.y = element_blank(),
              plot.margin = (unit(c(.5, .5, 2, 2), "cm"))) +
        # scale_x_continuous(breaks=c(seq(from = 0.8, to = 1.20, by = 0.1))) +
        #geom_text(aes(0.85,80, label="(a) r = 0.34, p = 0.027"),size=4) +
        xlab("cumulative substances from bootstrap") +
        ylab("mantelR") 
```

The plot peaks at 10 substances. We now want to do a single mantel test for chemical bray-curtis similarities based on these 10 compounds and genetic relatedness. As already shown in the plot, the mantelR is 0.164 and is highly significant.

```{r, warning=FALSE}
# indices of the 10 best compounds associated with relatedness -----------------
comp_ind_m <- c(36,52,86,88,96,103,110,203,206,207)

# bray curtis similarity matrix based on this 10 compounds
scent_bc <- 1-(as.matrix(vegan::vegdist(as.matrix(scent[factors$age == 1, comp_ind_m])),
                      method = "bray"))
# relatedness matrix
rel_m <- relatedness[1:41, 1:41]

# mantel test for association between both
vegan::mantel(rel_m, scent_bc, method = "spearman", permutation = 1000, na.rm = TRUE)
```

***

R version and platform.
```{r}
sessionInfo()
```